libvmi/os/freebsd/core.c

```diff
@@ -0,0 +1,258 @@
+        goto _exit;
+    }
+
+    dbprint(VMI_DEBUG_MISC, "**set vmi->kpgd (0x%.16"PRIx64").\n", vmi->kpgd);
+
+    os_interface = g_malloc(sizeof(struct os_interface));
+    if ( !os_interface )
+        goto _exit;
+
+    bzero(os_interface, sizeof(struct os_interface));
+    os_interface->os_get_offset = freebsd_get_offset;
+    os_interface->os_pid_to_pgd = freebsd_pid_to_pgd;
+    os_interface->os_pgd_to_pid = freebsd_pgd_to_pid;
+    os_interface->os_ksym2v = freebsd_symbol_to_address;
+    os_interface->os_usym2rva = NULL;
+    os_interface->os_v2sym = freebsd_system_map_address_to_symbol;
+    os_interface->os_read_unicode_struct = NULL;
+    os_interface->os_teardown = freebsd_teardown;
+
+    vmi->os_interface = os_interface;
+
+    return VMI_SUCCESS;
+
+_exit:
+    g_free(vmi->os_data);
+    vmi->os_data = NULL;
+    return VMI_FAILURE;
+}
+
+void freebsd_read_config_ghashtable_entries(char* key, gpointer value,
+        vmi_instance_t vmi)
+{
+
+    freebsd_instance_t freebsd_instance = vmi->os_data;
+
+    if (strncmp(key, "sysmap", CONFIG_STR_LENGTH) == 0) {
+        freebsd_instance->sysmap = strdup((char *)value);
+        goto _done;
+    }
+
+    if (strncmp(key, "rekall_profile", CONFIG_STR_LENGTH) == 0) {
+        freebsd_instance->rekall_profile = strdup((char *)value);
+        goto _done;
+    }
+
+    if (strncmp(key, "freebsd_pmap", CONFIG_STR_LENGTH) == 0) {
+        freebsd_instance->pmap_offset = *(addr_t *)value;
+        goto _done;
+    }
+
+    if (strncmp(key, "freebsd_vmspace", CONFIG_STR_LENGTH) == 0) {
+        freebsd_instance->vmspace_offset = *(addr_t *)value;
+        goto _done;
+    }
+
+    if (strncmp(key, "freebsd_pid", CONFIG_STR_LENGTH) == 0) {
+        freebsd_instance->pid_offset = *(addr_t *)value;
+        goto _done;
+    }
+
+    if (strncmp(key, "freebsd_name", CONFIG_STR_LENGTH) == 0) {
+        freebsd_instance->name_offset = *(addr_t *)value;
+        goto _done;
+    }
+
+    if (strncmp(key, "freebsd_pgd", CONFIG_STR_LENGTH) == 0) {
+        freebsd_instance->pgd_offset = *(addr_t *)value;
+        goto _done;
+    }
+
+    if (strncmp(key, "ostype", CONFIG_STR_LENGTH) == 0 || strncmp(key, "os_type", CONFIG_STR_LENGTH) == 0) {
+        goto _done;
+    }
+
+    if (strncmp(key, "name", CONFIG_STR_LENGTH) == 0) {
+        goto _done;
+    }
+
+    if (strncmp(key, "domid", CONFIG_STR_LENGTH) == 0) {
+        goto _done;
+    }
+
+    if (strncmp(key, "physoffset", CONFIG_STR_LENGTH) == 0) {
+        goto _done;
+    }
+
+    warnprint("Invalid offset %s given for FreeBSD target\n", key);
+
+_done:
+    return;
+}
+
+status_t freebsd_get_offset(vmi_instance_t vmi, const char* offset_name, addr_t *offset)
+{
+    const size_t max_length = 100;
+    freebsd_instance_t freebsd_instance = vmi->os_data;
+
+    if (freebsd_instance == NULL) {
+        errprint("VMI_ERROR: OS instance not initialized\n");
+        return 0;
+    }
+
```

Comment:
Minor nit: check offset_name and offset for NULL -- strncmp is not NULL tolerant and because you're dereferencing the offset parameter. Same comment applies to similar functions.